WIP: compiler modules

adriaanm opened the pull request:
Enforce consistency of pack, docs, osgi, maven tasks.

Instead of scattering properties all over invocations of
staged-docs, staged-pack, make-bundle, copy-bundle and mvn-package,
these are now all fully determined by the project they act on.

The varying arguments to these macros are all centralized.
They are named like `project.prop`, where `prop` is one of:
description, package, dir, name, namesuffix, version,
targetdir, targetjar, jar, docroot, skipPackage, srcdir, src, srcjar.

The defaults for these properties are computed using `init-project-prop`.

adriaanm committed:
Enforce consistency of pack, docs, osgi, maven tasks.

Instead of scattering properties all over invocations of
staged-docs, staged-pack, make-bundle, copy-bundle and mvn-package,
these are now all fully determined by the project they act on.

The varying arguments to these macros are all centralized.
They are named like `project.prop`, where `prop` is one of:
description, package, dir, name, namesuffix, version,
targetdir, targetjar, jar, docroot, skipPackage, srcdir, src, srcjar.

The defaults for these properties are computed using `init-project-prop`.

adriaanm committed:
De-duplicate logic in maven deployment.

adriaanm committed:
WIP: scala-compiler-doc and scala-compiler-interactive modules

Presentation does not depend on scaladoc (anymore/yet?)

Dbuild is getting there!

adriaanm commented:
the last commit is work in progress -- the first ones simplify the build to make the last one as small as possible

todo:
- somehow jline breaks in sbt console
- publish virtual (and how to convince the maven task to publish without a jar?) maven artifact scala-compiler that depends on -core, -doc, and -interactive? no more scala-compiler.jar? scala-compiler-*.jar instead?
- similarly for osgi

retronym commented:
> publish virtual (and how to convince the maven task to publish without a jar?) maven artifact scala-compiler that depends on -core, -doc, and -interactive? no more scala-compiler.jar? scala-compiler-*.jar instead?

From command line maven, I think it is enough to have a pom with `<packaging>pom</packaging>`. The `<maven>` ant task should have the same semantics, if you can't convince the more fine-grained ant tasks to do this.

retronym commented:
I guess you already know about this one, but here is an example: https://github.com/scala/scala-library-all

skyluc commented:
The last commit breaks the sbinary build https://scala-webapps.epfl.ch/jenkins/job/pr-scala-integrate-ide/3608/console

adriaanm commented:
Yep, I was trying to mimic scala-library-all, but the maven ant task does not let me publish without a jar.

adriaanm commented:
Thanks, @skyluc! I'm abusing jenkins to speed up my integration testing as I experiment with different poms/bnds for the scala-compiler split into scala-compiler-core, scala-compiler-doc, scala-compiler-interactive (and next up: scala-compiler-repl)

adriaanm commented:
Back to the original approach: slim down scala-compiler jar, moving those classes to newly introduced jars for scaladoc, presentation compiler.

adriaanm commented:
@skyluc, @dragos: could you have a look at the changes to the *.bnd files? thanks!

adriaanm commented:
btw, you can compile and publish this locally pretty quickly with:

```
ant -Dmaven.version.number=2.11.0-local -Dstarr.version=2.11.0-M6 -Dstarr.use.released=1 -Dlocker.skip=1 -Ddocs.skip=1 publish.local
```

dragos commented on src/build/bnd/scala-compiler.bnd:
```diff
@@ -6,3 +6,5 @@ Export-Package: *;version=${ver}
 Import-Package: jline.*;resolution:=optional, \
                 org.apache.tools.ant.*;resolution:=optional, \
                 *
+Require-Bundle: org.scala-lang.modules.scala-compiler-interactive_@SCALA_BINARY_VERSION@;bundle-version=@SCALA_COMPILER_INTERACTIVE_VERSION@, \
```
Why do you need these dependencies? I thought they go in the opposite direction.

adriaanm replied:
We don't need them -- I was thinking this way you wouldn't have to change the IDE build.

Ichoran committed:
Collections library tidying and deprecation.  Separate parts are listed below.

Collections library tidying, part one: scripting.

Everything in scala.collection.scripting is deprecated now, along with the
<< method that is implemented in a few other classes.  Scripting does not
seem used at all, and anyone who did can easily write a wrapper that does
the same thing.

Deprecated *Proxy collections.

The only place proxies were used in the library was in swing.ListView, and
that was easy to change to a lazy val.

Proxy itself is used in ScalaNumberProxy and such, so it was left
undeprecated.

Deprecated Synchronized* traits from collections.

Synchronizability does not compose well, and it requires careful examination
of every method (which has not actually been done).

Places where the Scala codebase needs to be fixed (eventually) include:
  scala.reflect.internal.util.Statistics$QuantMap
  scala.tools.nsc.interactive.Global (several places)

Deprecated LinkedList (including Double- and -Like variants).

Interface is idiosyncratic and dangerously low-level.  Although some
low-level functionality of this sort would be useful, this doesn't seem
to be the ideal implementation.

Also deprecated the extractFirst method in Queue as it exposes LinkedList.
Cannot shift internal representations away from LinkedList at this time
because of that method.

Deprecated non-finality of several toX collection methods.

Improved documentation of most toX collection methods to describe what the
expectation is for their behavior.  Additionally deprecated overriding of
  - toIterator in IterableLike (should always forward to iterator)
  - toTraversable in TraversableLike (should always return self)
  - toIndexedSeq in immutable.IndexedSeq (should always return self)
  - toMap in immutable.Map (should always return self)
  - toSet in immutable.Set (should always return self)

Did not do anything with IterableLike.toIterable or Seq/SeqLike.toSeq since
for some odd reason immutable.Range overrides those.

Deprecated Forwarders from collections.

Forwarding, without an automatic mechanism to keep up to date with changes
in the forwarded class, is inherently unreliable.  Absent a mechanism to
keep current, they're deprecated.  ListBuffer is the only class in the
collections library that uses forwarders, and that functionality can be
rolled into ListBuffer itself.

Deprecating immutable set/map adaptors.

They're a bad idea (barring compiler support) for the same reason that all
the other adaptors are a bad idea: they get out of date and probably have a
variety of performance bugs.

Deprecated inheritance from leaf classes in immutable collections.

Inheriting from leaf-classes in immutable collections is rarely a good idea
since whenever you use any interesting collections method you'll revert to
the original class.  Also, the methods are often designed to work with only
particular behavior, and an override would be difficult (at best) to make
work.  Fortunately, people seem to have realized this and there are few to
no cases of people extending PagedSeq and TreeSet and the like.

Note that in many cases the classes will become sealed not final.

Deprecated overriding of methods and inheritance from various mutable
collections.

Some mutable collections seem unsuited for overriding since to override
anything interesting you would need vast knowledge of internal data
structures and/or access to private methods.  These include
  - ArrayBuilder.ofX classes.
  - ArrayOps
  - Some methods of BitSet (moved others from private to protected final)
  - Some methods of HashTable and FlatHashTable
  - Some methods of HashMap and HashSet (esp += and -= which just forward)
  - Some methods of other maps and sets (LinkedHashX, ListMap, TreeSet)
  - PriorityQueue
  - UnrolledBuffer

This is a somewhat aggressive deprecation, the theory being better to try it
out now and back off if it's too much than not attempt the change and be
stuck with collections that can neither be safely inherited nor have
implementation details changed.

Note that I have made no changes--in this commit--which would cause
deprecation warnings in any of the Scala projects available on Maven (at
least as gathered by Adriaan).  There are deprecation warnings induced
within the library (esp. for classes/traits that should become static) and
the compiler.  I have not attempted to fix all the deprecations in the
compiler as some of them touch the IDE API (but these mostly involved
Synchronized which is inherently unsafe, so this should be fixed
eventually in coordination with the IDE code base(s)).

Updated test checks to include new deprecations.

Used a higher level implementation for messages in JavapClass.

adriaanm committed:
Render relevant properties to buildcharacter.properties

adriaanm committed:
Extra repository hook via `extra.repo.url` var

Previously defaulted to 'https://oss.sonatype.org/content/repositories/releases',
now must be set manually. PR validation uses it to configure the repository
that has the artifacts published during validation.

adriaanm committed:
More useful ant targets for faster pr validation

adriaanm committed:
IDE needs actors/swing/continuations publish for validation.

As soon as they are also modularized, we can go back to
only building/publishing the core (lib/reflect/compiler).

JamesIry committed:
Merge pull request #3103 from Ichoran/topic/tidy-collections

Topic/tidy collections

adriaanm commented:
Closing for rebasing and dbuild upgrade. Please pardon our dust.

JamesIry committed:
Merge pull request #3094 from retronym/topic/erasure-opt

Avoid needless TypeRef allocation during erasure.

adriaanm committed:
Merge pull request #3108 from adriaanm/faster-build

Faster PR validation

adriaanm committed:
Use 2.11.0-M6 for starr.
